### PR TITLE
[JavaSpring] Multipart causing HttpMediaTypeNotSupportedException when parameter is different than File or String

### PR DESCRIPTION
**Pull request by redamessoudi**
## PR checklist
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Provided a full/minimal spec to reproduce the issue
- [x] Validated the input using an OpenAPI validator
- [x] Tested with the latest master to confirm the issue still exists
- [x] Searched for related issues/PRs
- [x] Built the project and updated samples

## Description

Having an endpoint :
- Consuming `multipart/form-data`
- File as parameter
- Other parameters different than String (e.g Boolean, Long, Enums, ...)

The generated API crashes `HttpMediaTypeNotSupportedException` with error `Unsupported Media Type` and status `415` because of the usage of `@RequestPart` for parameters different than File or String.

```yaml
{
  "status": 415,
  "error": "Unsupported Media Type",
  "message": "Content type 'application/octet-stream' not supported"
}
```

Spring docs explained it [here](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestPart.html) as : 

> Note that @RequestParam annotation can also be used to associate the part of a "multipart/form-data" request with a method argument supporting the same method argument types. 

> The main difference is that when the method argument is not a String, @RequestParam relies on type conversion via a registered Converter or PropertyEditor while @RequestPart relies on HttpMessageConverters taking into consideration the 'Content-Type' header of the request part. @RequestParam is likely to be used with name-value form fields while @RequestPart is likely to be used with parts containing more complex content (e.g. JSON, XML).

## openapi-generator version
5.X.X and master

## OpenAPI example
```yaml
openapi: '3.0.1'
info:
  version: 1.0.0
  title: MultipartFile test
paths:
  /multipart-mixed:
    post:
      tags:
        - multipart
      description: Mixed MultipartFile test
      operationId: multipartMixed
      requestBody:
        content:
          multipart/form-data:
            schema:
              type: object
              required:
                - status
                - file
              properties:
                status:
                  $ref: '#/components/schemas/MultipartMixedStatus'
                marker:
                  description: "additional object"
                  type: object
                  properties:
                    name:
                      type: string
                file:
                  description: "a file"
                  type: string
                  format: binary
      responses:
        '204':
          description: Successful operation
components:
  schemas:
    MultipartMixedStatus:
      description: "additional field as Enum"
      type: string
      enum:
        - ALLOWED
        - IN_PROGRESS
        - REJECTED
      example: REJECTED
```
## Steps to reproduce
Generate **server** as a target using JavaSpring

## Generated code
```java
default ResponseEntity<Void> multipartMixed(
@ApiParam(value = "", required = true, allowableValues = "ALLOWED, IN_PROGRESS, REJECTED") @Valid 
@RequestPart(value = "status", required = true) MultipartMixedStatus status,
@ApiParam(value = "a file", required = true) 
@RequestPart(value = "file", required = true) MultipartFile file,
@ApiParam(value = "") @Valid 
@RequestPart(value = "marker", required = false) MultipartMixedMarker marker
    ) {
        return getDelegate().multipartMixed(status, file, marker);
    }
```

## Expected code
```java
default ResponseEntity<Void> multipartMixed(
@ApiParam(value = "", required = true, allowableValues = "ALLOWED, IN_PROGRESS, REJECTED") @Valid 
@RequestParam(value = "status", required = true) MultipartMixedStatus status,
@ApiParam(value = "a file", required = true) 
@RequestPart(value = "file", required = true) MultipartFile file,
@ApiParam(value = "") @Valid 
@RequestParam(value = "marker", required = false) MultipartMixedMarker marker
    ) {
        return getDelegate().multipartMixed(status, file, marker);
    }
```

## Solution
In JavaSpring module in `formParams.mustache`, I switched the parameter type to `@RequestParam` (instead of `@RequestPart`) as long as it is not a File.

### CODE REVIEW

**Comment by redamessoudi** [comment]
Cc: @wing328, @welshm, could you take a look at this.

**Comment by welshm** [comment]
Looks good to me

**Comment by redamessoudi** [comment]
cc @wing328 @welshm, could this be merged?

**Comment by welshm** [comment]
@cachescrubber was any of this captured in your recent changes?

**Comment by cachescrubber** [comment]
> @cachescrubber was any of this captured in your recent changes?

I don't think so. But I always wondered why  the spring-cloud library has it's own version of the **formParam.mustache** using @RequestParam instead of @RequestPart.

IMO we should really dig into this to fully understand the implication of changing this.

**Comment by cachescrubber** [comment]
Here `@RequestPart` is used to map `MultipartFile`: https://www.baeldung.com/sprint-boot-multipart-requests.

@redamessoudi Does your client uses the correct content type?

**Comment by redamessoudi** [comment]
> Here `@RequestPart` is used to map `MultipartFile`: https://www.baeldung.com/sprint-boot-multipart-requests.

With the fix of this PR, we continue to use  `@RequestPart` to map `MultipartFile`, only different parameters than File will use `@RequestParam`.


> @redamessoudi Does your client uses the correct content type?

Yes, my client uses the correct content type. Please refer to the given example in the PR description.

The [Spring documentation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestPart.html) should help.

**Comment by GeitV** [comment]
We're having the same issue in dart-generated (dart-dio-next) code. In our java project, we have defined OpenAPI request with file and request object in `multipart/form-data`, but our app team then tries to use generated code and gets
```
{timestamp: 2022-02-15T10:34:18.668+00:00, status: 415, error: Unsupported Media Type, message: Content type 'application/octet-stream' not supported, path: /really/cool/path}
```

**Comment by cachescrubber** [comment]
> With the fix of this PR, we continue to use `@RequestPart` to map `MultipartFile`, only different parameters than File will use `@RequestParam`.

@redamessoudi Ok, got it. According to the spring documentation this is indeed correct!

PR looks good.

Should we update the `spring-cloud/formaParams.mustache` accordingly? I never really understood why they are different. Was it an attempt so solve the exact same problem?

@wing328 @welshm The updated sample `src/test/resources/3_0/form-multipart-binary-array.yaml` does not generate a valid project (MultipartMixedStatus is not generated). Same with the example here in the issue. Could you check this?

**Comment by wing328** [comment]
@redamessoudi can you please resolve the merge conflicts when you've time?

**Comment by redamessoudi** [comment]
> Should we update the `spring-cloud/formaParams.mustache` accordingly? I never really understood why they are different. Was it an attempt so solve the exact same problem?

@cachescrubber I took a quick look at `spring-cloud/formaParams.mustache` and I think they faced the same issue. We can apply the same fix. Maybe someone from the `spring-cloud` can be involved before applying that.

**Comment by redamessoudi** [comment]
> @redamessoudi can you please resolve the merge conflicts when you've time?

@wing328 conflicts resolved ✅

**Comment by cachescrubber** [comment]
> I bet when you just simply delete the `spring-cloud/formaParams.mustache` we are all set.
> 
> Idea: Generate both spring-boot (server) and spring-cloud (client) to verify interoperability.

I just did the verification:

As-Is (using @RequestParam)

```
2022-02-18 19:13:50.551 DEBUG 76263 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] ---> POST http://localhost:8080/mixed/multipart-mixed?status=anfang&file=org.openapitools.api.TestInvocation%241%4067763ebe&marker=ende HTTP/1.1
2022-02-18 19:13:50.551 DEBUG 76263 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Content-Type: multipart/form-data
2022-02-18 19:13:50.552 DEBUG 76263 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] ---> END HTTP (0-byte body)
2022-02-18 19:13:50.690 DEBUG 76263 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] <--- HTTP/1.1 500 (135ms)
```

Using @**RequestPart**

```
2022-02-18 19:18:34.724 DEBUG 76325 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] ---> POST http://localhost:8080/mixed/multipart-mixed?status=anfang&marker=ende HTTP/1.1
2022-02-18 19:18:34.724 DEBUG 76325 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Content-Length: 167
2022-02-18 19:18:34.724 DEBUG 76325 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Content-Type: multipart/form-data; charset=UTF-8; boundary=17f0e0ea360
2022-02-18 19:18:34.725 DEBUG 76325 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] 
2022-02-18 19:18:34.725 DEBUG 76325 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Binary data
2022-02-18 19:18:34.725 DEBUG 76325 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] ---> END HTTP (167-byte body)
```

**Comment by cachescrubber** [comment]
> @cachescrubber I took a quick look at `spring-cloud/formaParams.mustache` and I think they faced the same issue. We can apply the same fix. Maybe someone from the `spring-cloud` can be involved before applying that.

@redamessoudi Could you apply the same fix to `spring-cloud/formaParams.mustache`?

**Comment by redamessoudi** [comment]
> @redamessoudi Could you apply the same fix to `spring-cloud/formaParams.mustache`?

Done.  Could you relaunch your test case?

**Comment by cachescrubber** [comment]
@redamessoudi could you please re generate the samples? `# ./bin/generate-samples.sh `

**Comment by redamessoudi** [comment]
@cachescrubber samples generated.

**Comment by cachescrubber** [comment]
Thanks. I just removed `spring-cloud/formaParams.mustache` locally and run the samples. The only differenece is beanvalidation (@Valid before @RequestParam) We use beanvalidation for other parameter types already, so this should not be no problem.

@redamessoudi could you remove `spring-cloud/formaParams.mustache` and re generate the samples?

File: samples/openapi3/server/petstore/springboot-implicitHeaders/src/main/java/org/openapitools/api/FakeApi.java [hunk 1]
```diff
@@ -357,7 +356,7 @@
         @Parameter(name = "number", description = "None", required = true) @Valid @RequestParam(value = "number", required = true) BigDecimal number,
         @Parameter(name = "double", description = "None", required = true) @Valid @RequestParam(value = "double", required = true) Double _double,
         @Parameter(name = "pattern_without_delimiter", description = "None", required = true) @Valid @RequestParam(value = "pattern_without_delimiter", required = true) String patternWithoutDelimiter,
-        @Parameter(name = "byte", description = "None", required = true) @Valid @RequestPart(value = "byte", required = true) byte[] _byte,
+        @Parameter(name = "byte", description = "None", required = true) @Valid @RequestParam(value = "byte", required = true) byte[] _byte,
```
Comment (cachescrubber) [inline samples/openapi3/server/petstore/springboot-implicitHeaders/src/main/java/org/openapitools/api/FakeApi.java:359]: Just spotted this change. What abut `byte[]`? In the spec this is `type: string, format: byte`  which is documented as `base64 encoded characters`.

Comment (redamessoudi) [reply samples/openapi3/server/petstore/springboot-implicitHeaders/src/main/java/org/openapitools/api/FakeApi.java:359]: I did some tests and seems not to break the current behavior. I can upload the byte array while the endpoint uses `@RequestParam`.
[`ByteArrayHttpMessageConverter`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/converter/ByteArrayHttpMessageConverter.html) is used for conversion.

**Comment by redamessoudi** [comment]
> Thanks. I just removed `spring-cloud/formaParams.mustache` locally and run the samples. The only differenece is beanvalidation (@Valid before @RequestParam) We use beanvalidation for other parameter types already, so this should not be no problem.
> 
> @redamessoudi could you remove `spring-cloud/formaParams.mustache` and re generate the samples?

@cachescrubber `spring-cloud/formParams.mustache` removed and samples regenerated. Sorry for the delay.

**Comment by cachescrubber** [comment]
@redamessoudi Thanks - I'll check the generated client and give feedback here.

**Comment by cachescrubber** [comment]
I just verified again by generating a client (spring-cloud) against server (spring-boot).

```
2022-02-26 09:09:19.710 DEBUG 79121 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] ---> POST http://localhost:8080/multipart-mixed?status=XXX&marker=zzz HTTP/1.1
2022-02-26 09:09:19.711 DEBUG 79121 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Content-Length: 172
2022-02-26 09:09:19.711 DEBUG 79121 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Content-Type: multipart/form-data; charset=UTF-8; boundary=17f3513ba55
2022-02-26 09:09:19.711 DEBUG 79121 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] 
2022-02-26 09:09:19.711 DEBUG 79121 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] Binary data
2022-02-26 09:09:19.712 DEBUG 79121 --- [           main] org.openapitools.api.MultipartApiClient  : [MultipartApiClient#multipartMixed] ---> END HTTP (172-byte body)
```

**Review by cachescrubber — APPROVED** [review]
This PR is ready to merge.

**Comment by osiox** [comment]
Spring server with `reactive: true` (spring webflux) is broken when including file part and json part together, most likely because of this change. `@RequestPart` worked correctly (5.4.0), while `@RequestParam` does not work (6.0.0).

The regression seems to be supported by Spring documentation.

Quoting spring documentation again, from PR description:

> (...) @RequestParam is likely to be used with name-value form fields while _**@RequestPart is likely to be used with parts containing more complex content (e.g. JSON, XML)**_.

And quoting [@RequestParam documentation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestParam.html):

> Supported for annotated handler methods in Spring MVC and Spring WebFlux as follows:
> 
> - In Spring MVC, "request parameters" map to query parameters, form data, and parts in multipart requests. This is because the Servlet API combines query parameters and form data into a single map called "parameters", and that includes automatic parsing of the request body.
> - _**In Spring WebFlux, "request parameters" map to query parameters only. To work with all 3, query, form data, and multipart data, you can use data binding to a command object annotated with [ModelAttribute](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/ModelAttribute.html).**_

With this in mind, I guess this is not the correct fix, at least for `reactive: true`. @RequestParam will handle only query parameters and will not handle form data.

Not sure if I will have the time to create issue with minimal reproducible example or (even less likely) to create PR with a fix, so for now I will just leave a comment here.

**Comment by bkratz** [comment]
> Spring server with `reactive: true` (spring webflux) is broken when including file part and json part together, most likely because of this change. `@RequestPart` worked correctly (5.4.0), while `@RequestParam` does not work (6.0.0).

We have the same issue, even without `reactive:true`

**Comment by jorgerod** [comment]
Hi
@osiox @bkratz @Ferioney I have the same problem, have you found a workaround?

Thank you

**Pull request merged**